File: api/src/handlers/tags.dart
import 'package:http/http.dart' as http;
import 'dart:convert';

class Tag {
  final int id;
  final String name;

  Tag({this.id, this.name});

  factory Tag.fromJson(Map<String, dynamic> json) {
    return Tag(
      id: json['id'],
      name: json['name'],
    );
  }
}

class TagService {
  static const String url = 'http://localhost:8000/tags';

  static Future<List<Tag>> getTags() async {
    final response = await http.get(url);
    if (response.statusCode == 200) {
      List jsonResponse = json.decode(response.body);
      return jsonResponse.map((tag) => new Tag.fromJson(tag)).toList();
    } else {
      throw Exception('Failed to load tags from API');
    }
  }

  static Future<Tag> createTag(Tag tag) async {
    final response = await http.post(url,
        headers: {"Content-Type": "application/json"},
        body: jsonEncode(<String, String>{
          'name': tag.name,
        }));
    if (response.statusCode == 201) {
      return Tag.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to create tag');
    }
  }

  static Future<Tag> updateTag(int id, Tag tag) async {
    final response = await http.put('$url/$id',
        headers: {"Content-Type": "application/json"},
        body: jsonEncode(<String, String>{
          'name': tag.name,
        }));
    if (response.statusCode == 200) {
      return Tag.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to update tag');
    }
  }

  static Future<void> deleteTag(int id) async {
    final http.Response response = await http.delete('$url/$id');
    if (response.statusCode != 200) {
      throw Exception('Failed to delete tag');
    }
  }

  static Future<List<Tag>> getTasksByTag(int id) async {
    final response = await http.get('$url/$id/tasks');
    if (response.statusCode == 200) {
      List jsonResponse = json.decode(response.body);
      return jsonResponse.map((tag) => new Tag.fromJson(tag)).toList();
    } else {
      throw Exception('Failed to load tasks by tag from API');
    }
  }
}

File: api/src/handlers/tags.rs
use actix_web::{web, HttpResponse, Responder};
use crate::models::Tag;
use crate::PoolType;

pub async fn get_tags(db: web::Data<PoolType>) -> impl Responder {
    let connection = db.get().unwrap();
    let result = diesel::sql_query("SELECT * FROM tags")
        .load::<Tag>(&connection)
        .expect("Failed to fetch tags");

    HttpResponse::Ok().json(result)
}

pub async fn create_tag(db: web::Data<PoolType>, tag: web::Json<Tag>) -> impl Responder {
    let connection = db.get().unwrap();
    let result = diesel::insert_into(tags::table)
        .values(&tag.into_inner())
        .execute(&connection)
        .expect("Failed to create tag");

    HttpResponse::Created().finish()
}

pub async fn update_tag(db: web::Data<PoolType>, tag_id: web::Path<i32>, tag: web::Json<Tag>) -> impl Responder {
    let connection = db.get().unwrap();
    let result = diesel::sql_query("UPDATE tags SET name = $1 WHERE id = $2")
        .bind(&tag.name)
        .bind(tag_id.into_inner())
        .execute(&connection)
        .expect("Failed to update tag");

    HttpResponse::Ok().finish()
}

pub async fn delete_tag(db: web::Data<PoolType>, tag_id: web::Path<i32>) -> impl Responder {
    let connection = db.get().unwrap();
    let result = diesel::sql_query("DELETE FROM tags WHERE id = $1")
        .bind(tag_id.into_inner())
        .execute(&connection)
        .expect("Failed to delete tag");

    HttpResponse::NoContent().finish()
}

pub async fn get_tasks_by_tag(db: web::Data<PoolType>, tag_id: web::Path<i32>) -> impl Responder {
    let connection = db.get().unwrap();
    let result = diesel::sql_query("SELECT * FROM tasks WHERE id IN (SELECT task_id FROM task_tags WHERE tag_id = $1)")
        .bind(tag_id.into_inner())
        .load::<Task>(&connection)
        .expect("Failed to fetch tasks");

    HttpResponse::Ok().json(result)
}
